src/Microsoft.Identity.Web/Resource/TokenAuthorizationRules.cs

```diff
@@ -0,0 +1,150 @@
+        /// <param name="claimName">Name (type) of the claim to verify.</param>
+        /// <param name="expectedValue">Expected value.</param>
+        /// <param name="messageWhenClaimsNotFound">Error message thrown when the claim is not found.</param>
+        /// <param name="messageWhenUnexpectedClaimValue">Error message thrown when claim value is not the expected one.</param>
+        private static void VerifyClaimEquals(
+            ClaimsPrincipal claimsPrincipal,
+            string claimName,
+            string expectedValue,
+            string messageWhenClaimsNotFound,
+            string messageWhenUnexpectedClaimValue)
+        {
+            string? claimValue = ClaimsPrincipalExtensions.GetClaimValue(claimsPrincipal, claimName);
+            if (claimValue == null)
+            {
+                throw new UnauthorizedAccessException(messageWhenClaimsNotFound);
+            }
+
+            if (string.Equals(claimValue, expectedValue, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new UnauthorizedAccessException(messageWhenUnexpectedClaimValue);
+            }
+        }
+
+        /// <summary>
+        /// Verify the value of a claim.
+        /// </summary>
+        /// <param name="claimsPrincipal">Claims.</param>
+        /// <param name="claimName">Name (type) of the claim to verify.</param>
+        /// <param name="expectedValue">Expected value.</param>
+        /// <param name="messageWhenClaimsNotFound">Error message thrown when the claim is not found.</param>
+        /// <param name="messageWhenUnexpectedClaimValue">Error message thrown when claim value is not the expected one.</param>
+        private static void VerifyClaimDoesNotExistOrDoesNotEquals(
+            ClaimsPrincipal claimsPrincipal,
+            string claimName,
+            string expectedValue,
+            string messageWhenClaimsNotFound,
+            string messageWhenUnexpectedClaimValue)
+        {
+            string? claimValue = ClaimsPrincipalExtensions.GetClaimValue(claimsPrincipal, claimName);
+            if (claimValue != null)
+            {
+                if (!string.Equals(claimValue, expectedValue, StringComparison.OrdinalIgnoreCase))
+                {
+                    throw new UnauthorizedAccessException(messageWhenUnexpectedClaimValue);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Verify the value of a claim.
+        /// </summary>
+        /// <param name="claimsPrincipal">Claims.</param>
+        /// <param name="possibleClaimNames">Names (types) of the claims to verify.</param>
+        /// <param name="acceptedValues">Expected values.</param>
+        /// <param name="messageWhenClaimsNotFound">Error message thrown when the claim is not found.</param>
+        /// <param name="messageWhenUnexpectedClaimValue">Error message thrown when claim value is not the expected one.</param>
+        private static void VerifyClaimIsInAcceptedValues(
```

Comment:
Won't this throw an exception when the claim _does_ contain an expected value?

Reply:
Fixed